File: app/lib/data/dto/event.dart
import 'package:app/data/dto/event_user.dart';
import 'package:app/data/dto/tag.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'event.freezed.dart';
part 'event.g.dart';

@freezed
class Event with _$Event {
  factory Event({
    @JsonKey(name: '_id') required String id,
    required String idCreator,
    required String name,
    required int? unreadMessages,
    required String? imageUrl,
    required String? description,
    required DateTime? date,
    required bool appearInSearch,
    required bool showAllMessage,
    required bool entryAfterAdminApproval,
    required List<String> idPinnedMessages,
    required List<String> usersWhoWantToJoin,
    required List<EventUser> users,
    required List<Tag> tags,
  }) = _Event;

  factory Event.fromJson(Map<String, dynamic> json) => _$EventFromJson(json);
}

File: server/models/event.js
import mongoose from 'mongoose'

const Event = new mongoose.Schema(
    {
        idCreator: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'User',
            required: true,
        },
        name: {
            type: String,
            index: true,
            required: true,
        },
        imageUrl: String,
        description: String,
        date: Date,
        appearInSearch: {
            type: Boolean,
            default: true,
        },
        showAllMessage: {
            type: Boolean,
            default: true,
        },
        entryAfterAdminApproval: {
            type: Boolean,
            default: false,
        },
        usersWhoWantToJoin: {
            type: Array,
            of: mongoose.Schema.Types.ObjectId,
            ref: 'User',
        },
        users: {
            type: Array,
            of: {
                user: {
                    type: mongoose.Schema.Types.ObjectId,
                    ref: 'User',
                    required: true,
                },
                date: {
                    type: Date,
                    default: Date.now,
                },
            },
            required: true,
        },
        tags: {
            type: Array,
            of: { type: mongoose.Schema.Types.ObjectId, ref: 'Tag' },
        },
    },
    {
        versionKey: false,
    },
)

export default mongoose.model('Event', Event)
